### PR TITLE
feat(parser): add paren-less single-param lambda `s => expr` (#1572)

### PR DESCRIPTION
**Pull request by t0k0sh1**
## Summary

Closes #1572.

Single-parameter lambdas may now omit the parentheses when the parameter has no type annotation and the body is a single expression: `xs.filter(s => s == "1")`. Multi-arg, type-annotated, and block-bodied lambdas keep the existing paren-required syntax.

| Input | Status |
|---|---|
| `s => expr` | ✅ new |
| `(s) => expr`, `(s: T) => expr`, `() => expr` | ✅ unchanged |
| `s, t => s + t` | ❌ rejected (multi-arg ambiguity with tuple destructure) |
| `s: T => expr` | ❌ rejected (typed-decl ambiguity) |
| `s =>\n  body` | ❌ rejected (block body needs parens) |
| `if cond => then else else` (bare-ident cond) | ✅ preserved |

### Implementation

- `parsePrimary`'s `Ident` branch dispatches a single-param lambda when `peek == FatArrow && !in_if_cond_`. Validates `isCamelCase`, builds `LambdaExpr` with one `any` param, parses body via `parseConditional()` with `in_async_fn_` save/restore (so async context never leaks into the lambda body).
- New `Parser::in_if_cond_` flag, set/restored by `parseIfExpression` around the cond parse, prevents the dispatch from preempting `if flag => then else else` (every `tests/spec/option_branch_merge_none.test.ry` site, etc.).
- Formatter is unchanged: existing `formatLambdaSig` already emits canonical `(s: any) => ...` so `s => x` round-trips to the paren form (case B from the plan).
- Knowledge: new `.claude/rules/parser-conventions.md` entry locks in the `in_if_cond_` invariant for future `Ident <suffix>` dispatches.

## Test plan

- [x] C++ tests: 9 new `BareLambda*` cases in `tests/test_parser.cpp` + 2 in `tests/test_formatter.cpp` (acceptance / multiple rejection branches / nested right-assoc / if-cond positive sibling / async-context isolation / canonical formatter output / idempotency)
- [x] Spec tests: `tests/spec/lambda_paren_omit.test.ry` end-to-end with `filter` / `map` / chained / binding / `if cond` regression
- [x] Full `./build/ry test -p` (169 files) — no regression on `option_branch_merge_none.test.ry` and friends
- [x] ASan + UBSan clean
- [x] libFuzzer 60 sec + bare-lambda seed corpus — no crash
- [x] Doc example `xs.filter(s => s == "1").len()` runs and prints `2`

## Summary by CodeRabbit

* **New Features**
  * Single-parameter lambdas can omit parentheses when the parameter has no type and the body is a single expression (e.g., xs.filter(s => s == "1")).

* **Documentation**
  * Updated lambda docs with shorthand examples, usage patterns, and a limitations section describing when parentheses remain required.

* **Formatter**
  * Formatter normalizes the shorthand to canonical parenthesized form (e.g., f = (s: any) => s + 1).

* **Parsing / Behavior**
  * Shorthand is disallowed where it would conflict with existing if-expression syntax to preserve expected semantics.

* **Tests**
  * Added parser and formatter tests covering the shorthand and its edge cases.

### CODE REVIEW

File: src/parser_expr.cpp [hunk 1]
```diff
@@ -161,7 +161,10 @@ std::vector<StmtNode> Parser::parseIfExpressionBranchBody() {
 // The leading `if` token is consumed here.
 ExprPtr Parser::parseIfExpression() {
     Token ifTok = lex_.next(); // consume 'if'
+    bool prev_in_if_cond = in_if_cond_;
+    in_if_cond_ = true;
     ExprPtr cond = parseConditional();
+    in_if_cond_ = prev_in_if_cond;
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/parser_expr.cpp:173]: Good catch — applied the try/catch + rethrow pattern at both new sites:

- `parseIfExpression()` around `parseConditional()` for the cond — restores `in_if_cond_` on throw
- bare-lambda dispatch in `parsePrimary()` around `parseConditional()` for the body — restores `in_async_fn_` on throw

Will be in the next push to this branch. The existing `parseParenLambdaExpr` (`src/parser_expr.cpp:1007-1040`) has the same save/restore pattern but I left it out of this PR since it predates #1572 — flagging it for a follow-up issue rather than expanding scope here.

Verified: `*BareLambda*` + `*Lambda*` + `*If*` C++ tests all pass (72/72), and `tests/spec/lambda_paren_omit.test.ry` 5/5 pass after the change.
